docs/proposals/llm-hyperparameter-optimization-api.md

```diff
@@ -0,0 +1,173 @@
+# HyperParameter Optimization API for LLM Fine-Tuning
+
+- [HyperParameter Optimization API for LLM Fine-Tuning](#hyperparameter-optimization-api-for-llm-fine-tuning)
+  * [Links](#links)
+  * [Motivation](#motivation)
+  * [Goal](#goal)
+  * [Design for API](#design-for-api)
+    + [Example](#example)
+  * [Implementation](#implementation)
+  * [Advanced Functionalities](#advanced-functionalities)
+
+## Links
+
+- [katib/issues#2291 (Tuning API in Katib for LLMs)](https://github.com/kubeflow/katib/issues/2291)
+
+## Motivation
+
+The rapid advancements and growing popularity of Large Language Models (LLMs) have driven an increased need for effective LLMOps in Kubernetes environments. To address this, we developed a [train API](https://www.kubeflow.org/docs/components/training/user-guides/fine-tuning/) within the Training Python SDK, simplifying the process of fine-tuning LLMs using distributed PyTorchJob workers. However, hyperparameter optimization remains a crucial yet labor-intensive task for enhancing model performance. Automating this tuning process through a dedicated API will facilitate efficient and scalable exploration of hyperparameters, ultimately improving model performance and reducing manual effort.
+
+## Goal
+
+Our goal is to develop a high-level API for tuning hyperparameters of LLMs that simplifies the process of hyperparameter optimization in Kubernetes. This API will seamlessly integrate with external platforms like HuggingFace and S3 for importing pretrained models and datasets. By specifying parameters for the training objective, trial configurations, and PyTorch worker configurations, the API will automate the creation of experiments and execution of trials. This abstraction of Kubernetes infrastructure complexities will enable data scientists to optimize hyperparameters efficiently and effectively.
+
+## Design for API
+
+![Design for API](images/design_api.jpg)
+
+```python
+import kubeflow.katib as katib
+from kubeflow.katib import KatibClient
+
+class KatibClient(object):
+	
+	def tune(
+		name: str, 
+		namespace: Optional[str] = None,
+		model_provider_parameters = None,
+		dataset_provider_parameters = None,
+		trainer_parameters = None,
+		algorithm_name: str = "random",
+		algorithm_settings: Union[dict, List[models.V1beta1AlgorithmSetting], None] = None,
+		objective_metric_name: str = "eval_accuracy",
+		additional_metric_names: List[str] = [],
+		objective_type: str = "maximize",
+		objective_goal: float = None,
+		max_trial_count: int = None,
+		parallel_trial_count: int = None,
+		max_failed_trial_count: int = None,
+		resources_per_trial: Union[dict, client.V1ResourceRequirements, None] = None,
+		pytorch_config=katib.PyTorchConfig(
+			num_workers: int = 1,
+			num_procs_per_worker: int = 1,
+		),
```

Comment:
Does does it sound to you @tenzen-y @johnugeorge @deepanker13 ?
Maybe we should also implement the following parameter to the PyTorchConfig: `resources_per_worker`, since we don't know how to distribute `resources_per_trial` resources for each PyTorchJob worker.

Reply:
I agree with you. I have replaced `resources_per_trial` with `resources_per_worker`, and moved it into `pytorch_config`.